lib/SILOptimizer/Mandatory/TFUtilities.cpp

```diff
@@ -1510,6 +1519,129 @@ SILInstruction *SILTensorOpInfo::canonicalizeOperands(
+std::pair<StringRef, SILTensorOpInfo::OperandClass>
+GraphOperationDecoder::decodeAttributeName(Identifier name) {
+  auto nameStr = name.str();
+  // Figure out what the suffix is (if any).
+  auto dollarLoc = nameStr.find('$');
+
+  auto opClass = OperandClass::Normal;
+  if (dollarLoc != StringRef::npos) {
+    auto suffix = nameStr.drop_front(dollarLoc+1);
+    opClass = SILTensorOpInfo::getOperandClass(suffix).getValue();
+  }
+
+  // Slice the suffix off the attribute name and add the decoded version.
+  return { nameStr.substr(0, dollarLoc), opClass };
+}
+
+
+//===----------------------------------------------------------------------===//
+// Device Partitioning Utilities
+//===----------------------------------------------------------------------===//
+
+/// Scan the specified function, looking for logic that configures the current
+/// graph.
+GraphGlobalConfiguration
+GraphGlobalConfiguration::getForFunction(SILFunction &fn) {
+  DeviceType deviceType = DeviceType::CPU;
+  bool isTPUInfeedEnabled = false;
+
+  SILInstruction *firstFound = nullptr;
+  SmallVector<SILInstruction *, 4> configureInsts;
+  for (auto &bb : fn)
+    for (auto &inst : bb) {
+      // Scan for the device configuration ops if present.
+      auto tfopInfo = SILTensorOpInfo::decode(&inst);
+      if (!tfopInfo) continue;
+      bool isConfigOp = tfopInfo->opName == "tfc.configureTPU" ||
+                        tfopInfo->opName == "tfc.configureGPU";
+      if (!isConfigOp) continue;
+
+      configureInsts.push_back(&inst);
+      // If we found one, make sure we don't have more than one.
+      if (firstFound) {
+        diagnose(fn.getASTContext(), inst.getLoc().getSourceLoc(),
+                 diag::tf_multiple_device);
+        diagnose(fn.getASTContext(), firstFound->getLoc().getSourceLoc(),
+                 diag::tf_multiple_device_prev);
+        continue;
+      }
+
+      // Otherwise, remember this one and decode it.
+      firstFound = &inst;
+
+      // Eventually we'll support multiple different configuration ops, so
+      // we recheck the opcode here.
+      if (tfopInfo->opName == "tfc.configureTPU") {
+        // Decode: tfc.configureTPU(isInfeedEnabled: bool)
+        deviceType = DeviceType::TPU;
+        auto infeedEnabled =
+          cast<IntegerLiteralInst>(tfopInfo->getAttrOperand(0));
+        isTPUInfeedEnabled = !infeedEnabled->getValue().isNullValue();
+      } else if (tfopInfo->opName == "tfc.configureGPU") {
+        deviceType = DeviceType::GPU;
+      } else {
+        llvm_unreachable("unknown device configuration op");
+      }
+    }
+
+
+  // If the program didn't specify, fall back to the command line option.
+  if (!firstFound) {
+    // At most one of these test-only flags should be set.
+    assert (!TFTargetTPU || !TFTargetGPU);
+    if (TFTargetTPU)
+      deviceType = DeviceType::TPU;
+    if (TFTargetGPU)
+      deviceType = DeviceType::GPU;
+  }
+
+  if (firstFound) {
```

Comment:
Why not use an `else` to follow the last statement?

Reply:
That was existing code, I just moved it, fixed though, thanks!